### PR TITLE
APEXCORE-457 Added web service security documentation

### PR DESCRIPTION
**Pull request by pramodin**
@davidyan74 please review

### CODE REVIEW

File: docs/security.md [hunk 1]
```diff
@@ -74,6 +74,32 @@ The snippet below shows the how the credentials can be specified in the configur
 
 The property `dt.authentication.principal` specifies the Kerberos user principal and `dt.authentication.keytab` specifies the absolute path to the keytab file for the user.
 
+### Webservices security
```
Comment (davidyan74) [inline docs/security.md]: "Web Services" should be two words

File: docs/security.md [hunk 2]
```diff
@@ -74,6 +74,32 @@ The snippet below shows the how the credentials can be specified in the configur
...
+
+Alongside every Apex application is an application master process running called Streaming Container Manager (STRAM). STRAM manages the application by handling the various control aspects of the application such as orchestrating the execution of the application on the cluster, playing a key role in scalability and fault tolerance, providing application insight by collecting statistics among other functionality.
+
+STRAM provides a web service interface to introspect the state of the application, its various components and to make dynamic changes to the applications. Some examples of supported functionality are getting resource usage and partition information of various operators, getting operator statistics and changing properties of running operators.
```
Comment (davidyan74) [inline docs/security.md]: "STRAM provides a web service interface to introspect the state of the application and its components, and also to make dynamic changes to the applications."

File: docs/security.md [hunk 3]
```diff
@@ -142,7 +168,15 @@ When operators are running there will be effective processing rate differences b
 
 Like STRAM, streaming containers also need to communicate with NameNode to use HDFS persistence for reasons such as saving the state of the operators. In secure mode they also use NameNode delegation tokens for authentication. These tokens are also seeded by STRAM for the streaming containers.
 
+#### Stram Webservices
```
Comment (davidyan74) [inline docs/security.md]: Web Services

File: docs/security.md [hunk 4]
```diff
@@ -142,7 +168,15 @@ When operators are running there will be effective processing rate differences b
...
+
+Clients connects to STRAM and make web service requests to obtain operational information about a runtime application. When security is enabled we want this connection to also be authenticated. In this mode the client passes a web service token in the request and the STRAM checks this token. If the token is valid, then the request is processed else it is denied.
```
Comment (davidyan74) [inline docs/security.md]: "Clients connect" or "A client connects ... and makes"

Comment (davidyan74) [inline docs/security.md]: the STRAM -> STRAM

File: docs/security.md [hunk 5]
```diff
@@ -142,7 +168,15 @@ When operators are running there will be effective processing rate differences b
...
+
+How does the client get the web service token in the first place The client will first have to first connect to STRAM via the Resource Manager Web Services Proxy which is a service run by Hadoop to proxy requests to application web services. This connection is authenticated by the proxy service using a protocol called SPNEGO when secure mode is enabled. SPNEGO is Kerberos over HTTP and the client also needs to support it. If the authentication is successful the proxy forwards the request to STRAM. STRAM in processing the request generates and send back a web service token similar to a delegation token. This token is then used by client in subsequent requests it makes directly to STRAM and STRAM is able to validate it since it generated the token in the first place.
```
Comment (davidyan74) [inline docs/security.md]: ... in the first place (add a question mark) The client ...

Comment (davidyan74) [inline docs/security.md]: "The client will first have to first connect ..."  Double "first"

Comment (davidyan74) [inline docs/security.md]: STRAM in processing the request generates and send**s** back

Comment (davidyan74) [inline docs/security.md]: This token is then used by (add **the**) client

**Comment by pramodin** [comment]
@davidyan74 Made the changes and rebased

**Pull request merged**